pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,1376 @@
+	user, err := t.auth.UserName(ctx)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+	if !userMatch(user, token) {
+		return apierrors.NewNotFound(GVR.GroupResource(), token.Name)
+	}
+
+	return t.SystemStore.Delete(token.Name, options)
+}
+
+func (t *SystemStore) Delete(name string, options *metav1.DeleteOptions) error {
+	err := t.secretClient.Delete(TokenNamespace, name, options)
+	if err == nil {
+		return nil
+	}
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	return apierrors.NewInternalError(fmt.Errorf("failed to delete token %s: %w", name, err))
+}
+
+func (t *Store) get(ctx context.Context, name string, options *metav1.GetOptions) (*ext.Token, error) {
+	sessionID := t.auth.SessionID(ctx)
+
+	// note: have to get token first before we can check for a user mismatch
+	token, err := t.SystemStore.Get(name, sessionID, options)
+	if err != nil {
+		return nil, err
+	}
+
+	user, err := t.auth.UserName(ctx)
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+	if !userMatch(user, token) {
+		return nil, apierrors.NewNotFound(GVR.GroupResource(), name)
+	}
+
+	return token, nil
+}
+
+func (t *SystemStore) Get(name, sessionID string, options *metav1.GetOptions) (*ext.Token, error) {
+	// Core token retrieval from backing secrets
+	// We try to go through the fast cache as much as we can.
+
+	var err error
+	var currentSecret *corev1.Secret
+	empty := metav1.GetOptions{}
+	if options == nil || *options == empty {
+		currentSecret, err = t.secretCache.Get(TokenNamespace, name)
+	} else {
+		currentSecret, err = t.secretClient.Get(TokenNamespace, name, *options)
+	}
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve token %s: %w", name, err))
+	}
+	token, err := tokenFromSecret(currentSecret)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to extract token %s: %w", name, err))
+	}
+
+	token.Status.Current = token.Name == sessionID
+	token.Status.TokenValue = ""
+	return token, nil
+}
+
+func (t *Store) list(ctx context.Context, options *metav1.ListOptions) (*ext.TokenList, error) {
+	user, err := t.auth.UserName(ctx)
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+
+	sessionID := t.auth.SessionID(ctx)
+
+	return t.SystemStore.list(false, user, sessionID, options)
+}
+
+func (t *SystemStore) List(options *metav1.ListOptions) (*ext.TokenList, error) {
+	return t.list(true, "", "", options)
+}
+
+func (t *SystemStore) list(fullView bool, user, sessionID string, options *metav1.ListOptions) (*ext.TokenList, error) {
+	// Core token listing from backing secrets
+	secrets, err := t.secretClient.List(TokenNamespace, *options)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to list tokens: %w", err))
+	}
+
+	var tokens []ext.Token
+	for _, secret := range secrets.Items {
```

Comment:
Is there any reason we're reading all tokens instead of using a selector to only ever list user's underlying secrets?

Reply:
Not really. ... Will work on getting something for this merged into the general `options` coming from the caller.

Reply:
Heh. I thought there was an impediment, when I checked the field selector support for Secrets.
Then I remembered that we are setting the UserID as a label on the backing secret precisely
because we want to be able to filter for it in spite of the limititations of kube filtering.

I will also have to see if `watch` can be made to use this as well, to pre-filter events.

Reply:
commit 754c8de5ca407a6fadb1af9d2fa3d9557796e3b8